File: scene_follow_evade.py
import asyncio

import util
from installation import Head


async def follow_closest_focal_point(
        head: Head,
        orchestrator: "Orchestrator",
        evade_distance: float = -1.0,
):
    while True:
        p = head.global_pos

        selected, distance = orchestrator.closest_focal_point_to(p)
        if selected is not None:
            # TODO: add some "hysteresis" here
            # TODO: random depending on time
            if distance < evade_distance:
                theta = head.point_away_from(selected)
            else:
                theta = head.point_to(selected)

            path = f"/rotation/{theta:f}"
            orchestrator.head_manager.send("head", head.name, path)

        await asyncio.sleep(0.04)


async def follow_evade(orchestrator: "Orchestrator"):
    tasks = []
    for head in orchestrator.inst.heads.values():
        task = util.create_task(
            follow_closest_focal_point(head, orchestrator, -1.0),
            allow_cancel=True,
        )
        tasks.append(task)
    await asyncio.gather(*tasks)

File: boss/scene_follow_evade.go
package boss

import (
	"context"
	"github.com/cacktopus/theheads/boss/scene"
	"github.com/cacktopus/theheads/boss/util"
	"github.com/cacktopus/theheads/boss/watchdog"
	gen "github.com/cacktopus/theheads/common/gen/go/heads"
	"github.com/sirupsen/logrus"
	"time"
)

const (
	trackingPeriod = 40 * time.Millisecond
)

func FollowClosestFocalPoint(
	dj *DJ,
	done util.BroadcastCloser,
	head *scene.Head,
	evadeDistance float64,
) {
	logger := logrus.WithField("head", head.Name)

loop:
	for {
		select {
		case <-time.After(trackingPeriod):
			p := head.GlobalPos()

			selected, distance := dj.grid.ClosestFocalPointTo(p)
			if selected == nil {
				continue
			}

			var theta float64
			if distance < evadeDistance {
				theta = head.PointAwayFrom(selected.Pos)
			} else {
				theta = head.PointTo(selected.Pos)
			}

			conn, err := dj.headManager.GetHeadConn(head.Name)
			if err != nil {
				logger.WithError(err).Error("error getting head conn")
				continue loop
			}

			if _, err = gen.NewHeadClient(conn).Rotation(context.Background(), &gen.RotationIn{
				Theta: theta,
			}); err != nil {
				logger.WithError(err).Error("error setting rotation")
				continue loop
			}
		case <-done.Chan():
			logger.Info("Finishing FollowClosestFocalPoint")
			return
		}
	}
}

func FollowEvade(dj *DJ, done util.BroadcastCloser, entry *logrus.Entry) {
	for _, head := range dj.scene.Heads {
		go FollowClosestFocalPoint(dj, done, head, -1.0)
	}

	t := time.NewTicker(5 * time.Second)

loop:
	for {
		select {
		case <-t.C:
			watchdog.Feed()
		case <-done.Chan():
			break loop
		}
	}

	logrus.Info("Finishing FollowEvade")
}
